hibernate-reactive/src/main/kotlin/com/linecorp/kotlinjdsl/query/HibernateMutinyReactiveQuery.kt

```diff
@@ -0,0 +1,70 @@
+package com.linecorp.kotlinjdsl.query
+
+import io.smallrye.mutiny.coroutines.awaitSuspending
+import org.hibernate.reactive.mutiny.Mutiny
+import javax.persistence.Parameter
+import kotlin.reflect.KClass
+import org.hibernate.reactive.session.ReactiveQuery as HibernateReactiveQuery
+
+class HibernateMutinyReactiveQuery<R>(private val query: Mutiny.Query<R>) : ReactiveQuery<R> {
+    override suspend fun singleResult(): R = query.singleResult.awaitSuspending()
+    override suspend fun resultList(): List<R> = query.resultList.awaitSuspending()
+    override suspend fun singleResultOrNull(): R? = query.singleResultOrNull.awaitSuspending()
+    override suspend fun executeUpdate(): Int = query.executeUpdate().awaitSuspending()
+
+    private val exception = UnsupportedOperationException("Hibernate-reactive does not support JPA query hint yet.")
+
+    override fun setParameter(position: Int, value: Any?): ReactiveQuery<R> {
+        query.setParameter(position, value)
+        return this
+    }
+
+    override fun setParameter(name: String, value: Any?): ReactiveQuery<R> {
+        query.setParameter(name, value)
+        return this
+    }
+
+    override fun <T> setParameter(parameter: Parameter<T>, value: T?): ReactiveQuery<R> {
+        query.setParameter(parameter, value)
+        return this
+    }
+
+    override fun setMaxResults(maxResults: Int): ReactiveQuery<R> {
+        query.maxResults = maxResults
+        return this
+    }
+
+    override fun setFirstResult(firstResult: Int): ReactiveQuery<R> {
+        query.firstResult = firstResult
+        return this
+    }
+
+    /**
+     * Hibernate Reactive's Mutiny.Query does not support setQueryHint method officially.
+     * but org.hibernate.reactive.session.ReactiveQuery has a setQueryHint method.
+     * Mutiny.Query (MutinyQueryImpl to be exact) has an org.hibernate.reactive.session.ReactiveQuery instance
+     * to delegate to ReactiveQuery methods.
+     * So we support setQueryHint by calling ReactiveQuery.setQueryHint Method included in Mutiny.Query.
+     * This reflection call will be removed in the future to officially support setQueryHint in Mutiny.Query.
+     * @see org.hibernate.reactive.session.ReactiveQuery#setQueryHint
+     */
+    override fun setQueryHint(hintName: String, value: Any) {
+        query::class.java.declaredFields
+            .find { it.type.name == HibernateReactiveQuery::class.qualifiedName }
+            ?.run {
+                isAccessible = true
+                (get(query) as HibernateReactiveQuery<*>).setQueryHint(hintName, value)
+                    ?: throw exception
+            } ?: throw exception
```

Comment:
-- english
This code is too dangerous, because it is a direct approach to the internal implementation of Hibernate.

-- korean
이 코드는 너무 위험한 코드 같습니다.

Hibernate의 내부 구현에 대해 직접 접근하는 것이기 때문에 너무 큰 위험부담을 우리가 가져가는 것으로 판단됩니다.

Reply:
This is not very dangerous code.
I have Hibernate's ReactiveQuery class and in that class it has a setQuernHint method explicitly. If the passed query object has the corresponding field, it will be executed normally, otherwise it will not be executed. Instead, an exception will be thrown. There is absolutely nothing hard-coded in this process.

-- korean
이건 그리 위험한 코드가 아닙니다.
하이버네이트의 ReactiveQuery 클래스가 있고 그 클래스에서 명확하게 setQuernHint 메소드를 가지고 있습니다. 넘겨 받은 쿼리 객체가 해당 필드를 가지고 있다면 정상 실행될 것이고 아니라면 실행되지 않을것입니다. 대신 예외가 발생하겠죠 이는 원래 우리가 생각했던 예외 발생 코드와 크게 다르지 않습니다. 이 과정에서 하드코딩된 부분도 전혀 없고요.

Reply:
Hibernate's ReactiveQuery cannot be considered as internal code, but is a representative interface of Stage and Mutiny Query, which are really internal implementations.
I tried to implement it because I thought that it would be much better to use it even if it is supported like this rather than not supporting it at all.

-- korean
Hibernate의 ReactiveQuery 는 내부 코드라고 볼수 없고 정말 내부 구현체인 Stage, Mutiny Query의 대표 인터페이스입니다.
아예 지원하지 않는 것보다는 이렇게라도 지원하는게 사용하는데 있어서 훨씬 낫지 않을까 싶어서 구현해 보았습니다.

Reply:
-- english
It's dangerous in that we don't use the public method provided by Mutiny.Query and MutinyQueryImpl, but we use the private method.

If MutinyQueryImpl decides to use another class without using ReactiveQuery, we will have to handle it. I think that the risk of maintenance is too high for us.

And since exceptions are generated by the creator, the stack trace can be taken strangely, so it would be right to make and throw exceptions inside.

-- korean
Mutiny.Query 및 MutinyQueryImpl에서 제공해주는 public 메소드를 쓰는 것이 아니라 private 메소드를 쓴다는 점에서 위험하다고 생각합니다.

만약 MutinyQueryImpl에서 ReactiveQuery를 delegate로 쓰지 않고 다른 클래스를 쓴다고 결정하면 저희는 그에 대응을 해주어야 할 것입니다. Hibernate의 내부 구현 내용까지 저희가 인터페이스 대상으로 삼는 것은 저희의 유지보수에 대한 위험 부담이 너무 크다고 생각됩니다.

그리고 예외를 생성자에서 생성하기 때문에 stack trace가 이상하게 찍힐 수 있어 내부에서 예외를 만들어서 던지는 것이 맞을 것 같습니다.

Reply:
I agree on that part.
Come to think of it, I think it might be because JPA hints are not fully supported. Even if it doesn't, we'll just support it according to the situation next time. Actually, I thought that not being able to comment was the biggest problem, but since SQL Comment is supported anyway, I can just do that.

i will fix it.

-- korean
그부분에 대해서는 공감합니다.
생각해보니 JPA 힌트에 대해 fully 지원하지 않아서 그럴수도 있겠다는 생각도 드네요. 혹 그렇지 않더라도 그냥 다음에 상황에 맞게 지원하도록 하도록 하겠습니다. 사실 comment 가 안되는게 제일 문제라고 생각했었는데 어차피 SQL Comment 는 지원하니 그냥 그렇게 해도 되겠네요.
수정하겠습니다.

Reply:
applied

-- korean
반영했습니다.